File: app/src/main/java/com/tunjid/fingergestures/activities/BrightnessActivity.java
package com.tunjid.fingergestures.activities;

import android.content.Intent;
import android.os.Bundle;
import android.transition.AutoTransition;
import android.transition.TransitionManager;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.gestureconsumers.BrightnessGestureConsumer;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;

import static android.graphics.PorterDuff.Mode.SRC_IN;
import static com.tunjid.fingergestures.gestureconsumers.BrightnessGestureConsumer.CURRENT_BRIGHTNESS_BYTE;

public class BrightnessActivity extends TimedActivity
        implements SeekBar.OnSeekBarChangeListener {

    private int brightnessByte;

    private SeekBar seekBar;
    private TextView seekBarText;
    private ViewGroup seekBarBackground;

    private BrightnessGestureConsumer brightnessGestureConsumer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_brightness);

        brightnessGestureConsumer = BrightnessGestureConsumer.getInstance();

        int sliderColor = backgroundManager.getSliderColor();
        int sliderBackgroundColor = backgroundManager.getBackgroundColor();

        ConstraintLayout layout = findViewById(R.id.constraint_layout);
        seekBarBackground = findViewById(R.id.seekbar_background);
        seekBar = findViewById(R.id.seekbar);
        seekBarText = findViewById(R.id.seekbar_text);
        ImageView settingsIcon = findViewById(R.id.go_to_settings);

        seekBarText.setTextColor(sliderColor);
        seekBar.getThumb().setColorFilter(sliderColor, SRC_IN);
        seekBar.getProgressDrawable().setColorFilter(sliderColor, SRC_IN);
        settingsIcon.setImageDrawable(backgroundManager.tint(R.drawable.ic_settings_white_24dp, sliderColor));
        seekBarBackground.setBackground(backgroundManager.tint(R.drawable.color_indicator, sliderBackgroundColor));

        ConstraintSet set = new ConstraintSet();
        set.clone(layout);
        set.setVerticalBias(seekBarBackground.getId(), brightnessGestureConsumer.getPositionPercentage() / 100F);
        set.applyTo(layout);

        layout.setOnClickListener(v -> finish());
        settingsIcon.setOnClickListener(v -> startActivity(new Intent(this, MainActivity.class)));

        handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleIntent(intent);
    }

    protected void onResume() {
        super.onResume();
        if (brightnessGestureConsumer.shouldAnimateSlider())
            overridePendingTransition(R.anim.slide_in_down, R.anim.slide_out_up);
    }

    @Override
    public void finish() {
        super.finish();

        boolean animateSlide = brightnessGestureConsumer.shouldAnimateSlider();
        if (animateSlide) overridePendingTransition(R.anim.slide_in_down, R.anim.slide_out_down);
        else overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int percentage, boolean fromUser) {
        updateEndTime();

        if (!fromUser) return;
        if (percentage == 100) percentage--;

        brightnessByte = brightnessGestureConsumer.percentToByte(percentage);
        brightnessGestureConsumer.saveBrightness(brightnessByte);

        if (seekBarText.getVisibility() != View.VISIBLE) return;

        TransitionManager.beginDelayedTransition(seekBarBackground, new AutoTransition());
        seekBarText.setVisibility(View.GONE);
    }

    private void handleIntent(Intent intent) {
        TransitionManager.beginDelayedTransition(seekBarBackground, new AutoTransition());

        brightnessByte = intent.getIntExtra(CURRENT_BRIGHTNESS_BYTE, 0);
        int percentage = brightnessGestureConsumer.byteToPercentage(brightnessByte);

        seekBar.setProgress(percentage, true);
        seekBar.setOnSeekBarChangeListener(this);

        boolean showDimmer = brightnessGestureConsumer.shouldShowDimmer();
        seekBarText.setVisibility(showDimmer ? View.VISIBLE : View.GONE);

        if (showDimmer) {
            float dimmerPercent = brightnessGestureConsumer.getScreenDimmerDimPercent() * 100F;
            seekBarText.setText(getString(R.string.screen_dimmer_value, dimmerPercent));
        }

        waitToFinish();
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        brightnessGestureConsumer.removeDimmer();
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) { }
}

File: app/src/main/java/com/tunjid/fingergestures/activities/BrightnessActivity.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.activities

import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.transition.AutoTransition
import android.transition.TransitionManager
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.view.isVisible
import androidx.dynamicanimation.animation.SpringForce
import androidx.dynamicanimation.animation.springAnimationOf
import com.google.android.material.slider.LabelFormatter
import com.google.android.material.slider.Slider
import com.tunjid.androidx.core.content.drawableAt
import com.tunjid.androidx.core.graphics.drawable.withTint
import com.tunjid.fingergestures.R
import com.tunjid.fingergestures.activities.main.activeOnCreateLifecycleOwner
import com.tunjid.fingergestures.databinding.ActivityBrightnessBinding
import com.tunjid.fingergestures.di.viewModelFactory
import com.tunjid.fingergestures.filter
import com.tunjid.fingergestures.gestureconsumers.BrightnessGestureConsumer.Companion.CURRENT_BRIGHTNESS_BYTE
import com.tunjid.fingergestures.map
import com.tunjid.fingergestures.mapDistinct
import com.tunjid.fingergestures.viewmodels.BrightnessViewModel
import com.tunjid.fingergestures.viewmodels.In
import com.tunjid.fingergestures.viewmodels.State

class BrightnessActivity : AppCompatActivity() {
    private val viewModel by viewModelFactory<BrightnessViewModel>()
    private val binding by lazy { ActivityBrightnessBinding.inflate(layoutInflater) }
    private val dialogLifecycleOwner = activeOnCreateLifecycleOwner()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        window.apply {
            window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            window.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        }

        val controls = binding.controls
        val sliderSpring = springAnimationOf(
            controls.slider::setValue,
            controls.slider::getValue
        ).setSpring(SpringForce(100f)
            .setDampingRatio(SpringForce.DAMPING_RATIO_LOW_BOUNCY)
        )

        controls.slider.labelBehavior = LabelFormatter.LABEL_GONE
        controls.slider.addOnChangeListener { _, value, fromUser ->
            if (fromUser) viewModel.accept(In.Change(value.toInt()))
        }
        controls.slider.addOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) = viewModel.accept(In.RemoveDimmer)

            override fun onStopTrackingTouch(slider: Slider) = Unit
        })

        viewModel.state.apply {
            mapDistinct(State::sliderColor)
                .observe(dialogLifecycleOwner) { sliderColor ->
                    controls.slider.thumbTintList = ColorStateList.valueOf(sliderColor)
                    controls.slider.trackTintList = ColorStateList.valueOf(sliderColor)
                    controls.goToSettings.setImageDrawable(drawableAt(R.drawable.ic_settings_white_24dp)
                        ?.withTint(sliderColor))
                }
            mapDistinct(State::backgroundColor)
                .observe(dialogLifecycleOwner) { backgroundColor ->
                    controls.sliderBackground.background = drawableAt(R.drawable.color_indicator)
                        ?.withTint(backgroundColor)
                }
            mapDistinct(State::verticalBias)
                .observe(dialogLifecycleOwner) { verticalBias ->
                    val layout = binding.constraintLayout

                    val set = ConstraintSet()
                    set.clone(layout)
                    set.setVerticalBias(controls.sliderBackground.id, verticalBias)
                    set.applyTo(layout)
                }
            mapDistinct(State::showDimmerText)
                .observe(dialogLifecycleOwner, {
                    TransitionManager.beginDelayedTransition(binding.controls.sliderBackground, AutoTransition())
                    controls.sliderText.isVisible = it
                })

            filter(State::showDimmerText)
                .mapDistinct(State::dimmerPercent)
                .observe(dialogLifecycleOwner) {
                    binding.controls.sliderText.text = getString(R.string.screen_dimmer_value, it)
                }
            mapDistinct(State::initialProgress)
                .map(Int::toFloat)
                .observe(dialogLifecycleOwner, sliderSpring::animateToFinalPosition)

            filter(State::completed)
                .observe(dialogLifecycleOwner) { finish() }
        }

        binding.constraintLayout.setOnClickListener { finish() }
        controls.goToSettings.setOnClickListener { startActivity(Intent(this, MainActivity::class.java)) }

        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    override fun onResume() {
        super.onResume()
        if (viewModel.state.value?.animateSlide == true)
            overridePendingTransition(R.anim.slide_in_down, R.anim.slide_out_up)
    }

    override fun finish() {
        super.finish()
        when (viewModel.state.value?.animateSlide) {
            true -> overridePendingTransition(R.anim.slide_in_down, R.anim.slide_out_down)
            else -> overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        }
    }

    private fun handleIntent(intent: Intent) {
        val brightnessByte = intent.getIntExtra(CURRENT_BRIGHTNESS_BYTE, 0)
        viewModel.accept(In.Start(brightnessByte))
    }
}
